src/Dialog/DialogBottomSheet.tsx

```diff
@@ -0,0 +1,318 @@
+  const startHeight = useRef(0)
+  const isDragging = useRef(false)
+
+  useRefObjectAsForwardedRef(forwardedRef, dialogRef)
+
+  // 🧑‍🦽 ACCESSIBILITY
+
+  const isReduced = useMedia('(prefers-reduced-motion: no-preference)', false) === false
+
+  // 🥊 ACTIONS
+
+  const showBottomSheet = () => {
+    setIsOpen(true)
+    updateSheetHeight(HALF_HEIGHT)
+  }
+
+  const hideBottomSheet = async () => {
+    setIsOpen(false)
+    if (isReduced) {
+      onClose()
+    } else {
+      setFireDelayedOnClose(true)
+    }
+  }
+
+  const updateSheetHeight = (height: number) => {
+    if (!dialogRef.current) return
+    dialogRef.current.style.height = `${height}dvh`
+    setSnappedHeight(height)
+  }
+
+  // 🎪 EVENTS
+
+  const dragStop = (e: MouseEvent | TouchEvent) => {
+    if (!dialogRef.current) return
+
+    e.preventDefault()
+
+    isDragging.current = false
+    const sheetHeight = parseInt(dialogRef.current.style.height)
+
+    dialogRef.current.style.transition = isReduced ? 'none' : '0.3s ease'
+
+    if (sheetHeight < 10) return hideBottomSheet()
+    if (sheetHeight > 75) return updateSheetHeight(FULL_HEIGHT)
+
+    updateSheetHeight(HALF_HEIGHT)
+  }
+
+  const dragStart = (e: MouseEvent | TouchEvent) => {
+    if (!dialogRef.current) return
+
+    e.preventDefault()
+
+    if (e.type === 'touchstart' && 'touches' in e) {
+      startY.current = e.touches[0].pageY
+    } else if ('clientX' in e) {
+      startY.current = e.pageY
+    }
+
+    startHeight.current = parseInt(dialogRef.current.style.height)
+    isDragging.current = true
+    dialogRef.current.style.transition = 'none'
+  }
+
+  const dragging = (e: MouseEvent | TouchEvent) => {
+    if (!dialogRef.current || !isDragging.current) return
+
+    e.preventDefault()
+
+    let pageY
+    if (e.type === 'touchmove' && 'touches' in e) {
+      pageY = e.touches[0].pageY
+    } else if ('clientX' in e) {
+      pageY = e.pageY
+    }
+
+    const delta = startY.current - (pageY || 0)
+    const newHeight = startHeight.current + (delta / window.innerHeight) * 100
+    updateSheetHeight(newHeight)
+  }
+
+  const onSliderChange = (e: ChangeEvent<HTMLInputElement>) => {
+    if (!isDragging.current) {
+      const value = parseInt(e.target.value)
+      if (value === 1) {
+        updateSheetHeight(HALF_HEIGHT)
+      } else if (value === 2) {
+        updateSheetHeight(FULL_HEIGHT)
+      }
+    }
+  }
+
+  // 🪝 HOOKS
+
+  // Ensures an animation upon closing the component
+  useEffect(() => {
+    if (!fireDelayedOnClose) return
+    setIsOpen(false)
+    const timer = setTimeout(() => {
+      onClose()
+    }, ANIMATION_DURATION)
+    return () => clearTimeout(timer)
+  }, [fireDelayedOnClose, onClose])
+
+  // Animates the bottom sheet in on mount
+  useEffect(() => {
+    showBottomSheet()
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  const currentSliderValue = snappedHeight === HALF_HEIGHT ? 1 : 2
+
+  return (
+    <FullScreenContainer
```

Comment:
Why does this container exist?

Reply:
It doesn't have to, but there are subtle differences that cause them to behave differently. When animation is used in the code, it can be easier to understand if you organize the different states by their names. Of course, this is a matter of personal preference, and you could also consider adding an 'isFullScreen' boolean property here.

``` JavaScript
const NormalDialog = styled.div<StyledDialogProps>`
  ...
  box-shadow: ${get('shadows.overlay.shadow')};
  min-width: 296px;
  max-width: calc(100vw - 64px);
  max-height: calc(100vh - 64px);
  ...
  width: ${props => widthMap[props.width ?? ('xlarge' as const)]};
  height: ${props => heightMap[props.height ?? ('auto' as const)]};
  border-radius: 12px;
  ...
  @keyframes overlay--dialog-appear {
    0% {
      opacity: 0;
      transform: scale(0.5);
    }
    100% {
      opacity: 1;
      transform: scale(1);
    }
  }

  ${sx};
`

const FullScreenDialog = styled.div<StyledDialogProps>`
  ...
  width: 100vw;
  height: 100vh;
  ...
  animation: overlay--dialog-appear ${ANIMATION_DURATION} ${get('animation.easeOutCubic')};
  @keyframes overlay--dialog-appear {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }
  ${sx};
`
```

Reply:
Same answer as: https://github.com/primer/react/pull/4048#discussion_r1455511524